### ISSUE TITLE
Reproducing MF-it result in Table 2 of the paper

### ISSUE THREAD

**Issue by HashmatShadab** [yunqing-me/AttackVLM#10]
```python
import argparse
import os
import random
import clip
import numpy as np
import torch
import torchvision
from PIL import Image
from imagenet_filtered import FilteredImageNetDataset
from torchvision.transforms.functional import normalize
from lavis.models import load_model_and_preprocess

# basic random seed
def seedBasic(seed):
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)


# torch random seed
def seedTorch(seed):
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False


# combine
def seedEverything(seed):
    seedBasic(seed)
    seedTorch(seed)


# ------------------------------------------------------------------ #

def to_tensor(pic):
    mode_to_nptype = {"I": np.int32, "I;16": np.int16, "F": np.float32}
    img = torch.from_numpy(np.array(pic, mode_to_nptype.get(pic.mode, np.uint8), copy=True))
    img = img.view(pic.size[1], pic.size[0], len(pic.getbands()))
    img = img.permute((2, 0, 1)).contiguous()
    return img.to(dtype=torch.get_default_dtype())



class GetTargetCaptions(torchvision.datasets.VisionDataset):

    def __init__(self, path):
        self.path = path
        # read txt file
        with open(self.path, 'r') as f:
            self.captions = f.readlines()
            # save in a list
            self.captions = [x.strip() for x in self.captions]

    def __len__(self):
        return len(self.captions)
    def __getitem__(self, index: int):
        prompts = self.captions[index]

        return prompts


def get_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--batch_size", default=1, type=int)
    parser.add_argument("--input_res", default=224, type=int)
    parser.add_argument("--clip_encoder", default="RN50", type=str)
    parser.add_argument("--alpha", default=1.0, type=float)
    parser.add_argument("--epsilon", default=8, type=int)
    parser.add_argument("--steps", default=100, type=int)
    parser.add_argument("--output", default=r".\mf_it_adv", type=str, help='the folder name of output')
    parser.add_argument("--cle_data_path", default=r"D:\Code\datasets\imagenet\val", type=str, help='path of the clean images')
    parser.add_argument("--image_json", type=str, default=r"image_list.json",
                        help='path of the clean images')
    parser.add_argument("--caption_path", type=str, default=None,
                        help='path of the clean images')
    parser.add_argument("--captions_per_image", default=1, type=int)
    parser.add_argument("--tgt_data_path", default=r"D:\Code\Projects\AttackVLM-main\stable-diffusion\coco_captions.txt",
                        type=str, help='path of the target captions')

    parser.add_argument("--model_name", default="blip2_opt", type=str)
    parser.add_argument("--model_type", default="pretrain_opt2.7b", type=str)

    args = parser.parse_args()
    return args


if __name__ == "__main__":

    args = get_args()

    # seed for everything
    # credit: https://www.kaggle.com/code/rhythmcam/random-seed-everything
    DEFAULT_RANDOM_SEED = 2023
    device = "cuda" if torch.cuda.is_available() else "cpu"
    seedEverything(seed=DEFAULT_RANDOM_SEED)

    # load clip_model params
    alpha = args.alpha
    epsilon = args.epsilon
    clip_model, preprocess = clip.load(args.clip_encoder, device=device)
    preprocess.transforms.pop()

    print(f"Loading LAVIS models: {args.model_name}, model_type: {args.model_type}...")
    model, vis_processors, txt_processors = load_model_and_preprocess(name=args.model_name, model_type=args.model_type, is_eval=True,
                                                                      device=device)
    print("Done.")

    # ------------- pre-processing images/text ------------- #

    clean_data = FilteredImageNetDataset(args.cle_data_path, image_json_path=args.image_json,
                                         caption_root_path=args.caption_path, transform=preprocess,
                                         num_caption_per_image=args.captions_per_image)

    target_data = GetTargetCaptions(args.tgt_data_path)

    data_loader_imagenet = torch.utils.data.DataLoader(clean_data, batch_size=args.batch_size, shuffle=False,
                                                       drop_last=False)
    data_loader_target = torch.utils.data.DataLoader(target_data, batch_size=args.batch_size, shuffle=False,
                                                     drop_last=False)

    clip_mean = (0.48145466, 0.4578275, 0.40821073)
    clip_std = (0.26862954, 0.26130258, 0.27577711)
    # CLIP imgs mean and std.

    # create a time variable to compute the time taken for the attack
    import time
    start_time = time.time()


    avg_clip_score_it = 0
    avg_clip_score_adv_it = 0
    avg_clip_score_tt = 0
    avg_clip_score_adv_tt = 0
    num_images = 0

    for i, ((image_org, label, caption, path), (prompts_tgt)) in enumerate(zip(data_loader_imagenet, data_loader_target)):

        num_images += len(image_org)
        image_org = image_org.to(device)
        prompts_tgt = prompts_tgt[0:len(image_org)]

        f_names = [os.path.split(path[i])[1] for i in range(len(path))]
        # save the path and its corresponding target caption

        # get tgt featutres
        with torch.no_grad():
            texts = clip.tokenize(prompts_tgt).to(device)
            txt_embedding_features = clip_model.encode_text(texts)
            txt_embedding_features /= txt_embedding_features.norm(dim=1, keepdim=True)

        # -------- get adv image -------- #
        adv_image = image_org.clone().detach()

        for j in range(args.steps):

            adv_image.requires_grad = True
            adv_image_features = clip_model.encode_image(normalize(adv_image, mean=clip_mean, std=clip_std))
            adv_image_features = adv_image_features / adv_image_features.norm(dim=1, keepdim=True)

            embedding_sim = torch.mean(
                torch.sum(adv_image_features * txt_embedding_features, dim=1))  # computed from normalized features (therefore it is cos sim.)

            grad = torch.autograd.grad(embedding_sim, adv_image, retain_graph=False, create_graph=False)[0]
            adv_image = adv_image.detach() + alpha/255.0 * torch.sign(grad.detach())
            delta = torch.clamp(adv_image - image_org, min=-epsilon/255.0, max=epsilon/255.0)
            adv_image = torch.clamp(image_org + delta, min=0.0, max=1.0)

            print(
                f"iter {i}/{len(data_loader_imagenet)} step:{j:3d}, embedding similarity={embedding_sim.item():.5f}, max delta={torch.max(torch.abs(delta)).item():.3f}, mean delta={torch.mean(torch.abs(delta)).item():.3f}")

        # compute the similarity between the adv image and the target caption
        with torch.no_grad():
            # adv_image = clip_preprocess(adv_image)
            adv_image_features = clip_model.encode_image(normalize(adv_image, mean=clip_mean, std=clip_std))
            adv_image_features = adv_image_features / adv_image_features.norm(dim=1, keepdim=True)
            embedding_sim = torch.mean(
                torch.sum(adv_image_features * txt_embedding_features, dim=1))
            print(f"Similarity between adv image and target caption: {embedding_sim.item():.5f}", '\n')
            avg_clip_score_adv_it += embedding_sim.item()*len(image_org)

        # compute the similarity between the original image and the target caption
        with torch.no_grad():
            # image_org = clip_preprocess(image_org)
            image_features = clip_model.encode_image(normalize(image_org, mean=clip_mean, std=clip_std))
            image_features = image_features / image_features.norm(dim=1, keepdim=True)
            embedding_sim = torch.mean(
                torch.sum(image_features * txt_embedding_features, dim=1))
            print(f"Similarity between original image and target caption: {embedding_sim.item():.5f}", '\n')
            avg_clip_score_it += embedding_sim.item()*len(image_org)

        # compute average clip score till batch i
        print(f"Batch {i}/{len(data_loader_imagenet)} Average CLIP Similarity between adv_image and tgt_text: {avg_clip_score_adv_it/num_images:.5f}", '\n')
        print(f"Batch {i}/{len(data_loader_imagenet)} Average CLIP Similarity between original image and tgt_text: {avg_clip_score_it/num_images:.5f}", '\n')


        # Get BLIP captions for the original and adv images
        with torch.no_grad():
            clean_samples = {"image": image_org}
            adv_samples = {"image": adv_image}
            clean_samples["image"] = normalize(clean_samples["image"], mean=clip_mean, std=clip_std)
            adv_samples["image"] = normalize(adv_samples["image"], mean=clip_mean, std=clip_std)
            clean_caption = model.generate(clean_samples, use_nucleus_sampling=True, num_captions=args.captions_per_image)
            adv_caption = model.generate(adv_samples, use_nucleus_sampling=True, num_captions=args.captions_per_image)

        # Get the txt embedding features for the BLIP captions
        with torch.no_grad():
            clean_text_token = clip.tokenize(clean_caption).to(device)
            clean_text_embedding_features = clip_model.encode_text(clean_text_token)
            clean_text_embedding_features = (clean_text_embedding_features / clean_text_embedding_features.norm(dim=1, keepdim=True)).detach()

            adv_text_token = clip.tokenize(adv_caption).to(device)
            adv_text_embedding_features = clip_model.encode_text(adv_text_token)
            adv_text_embedding_features = (adv_text_embedding_features / adv_text_embedding_features.norm(dim=1, keepdim=True)).detach()

        # compute the similarity between the txt_features from clean and adv with the target caption
        with torch.no_grad():
            txt_embedding_sim_clean = torch.mean(torch.sum(clean_text_embedding_features * txt_embedding_features, dim=1))
            print(f"Similarity between clean text and target caption: {embedding_sim.item():.5f}", '\n')
            avg_clip_score_tt += txt_embedding_sim_clean.item()*len(image_org)
            txt_embedding_sim_adv = torch.mean(torch.sum(adv_text_embedding_features * txt_embedding_features, dim=1))
            print(f"Similarity between adv text and target caption: {embedding_sim.item():.5f}", '\n')
            avg_clip_score_adv_tt += txt_embedding_sim_adv.item()*len(image_org)

     

    # save the avg clip score
    with open(os.path.join(args.output, 'avg_clip_score.txt'), 'a') as f:
        f.write("Average CLIP Score between adv_image and tgt_text: %s\n" % str(avg_clip_score_adv_it/num_images))
        f.write("Average CLIP Score between original image and tgt_text: %s\n" % str(avg_clip_score_it/num_images))
        f.write("Average CLIP Score between adv_text and tgt_text: %s\n" % str(avg_clip_score_adv_tt/num_images))
        f.write("Average CLIP Score between clean_text and tgt_text: %s\n" % str(avg_clip_score_tt/num_images))


    print("--- %s seconds ---" % (time.time() - start_time))

``` 

This is my code for reproducing the results for MF-it. I am using CLIP encoder as a surrogate model and maximizing the similarity between ImageNet samples and the corresponding coco captions. During the attack the similarity loss increases and I am able to get similar white box attack results as reported in Table 1. However, in the same loop when I pass the adversarial image through BLIP-2 model to compute the similarity between target text emebdding and the text emebdding generated on the adversarial image, the results don't match to the ones reported in Table 2 of the paper.

**Comment by HashmatShadab** [comment]
<img width="244" alt="image" src="https://github.com/yunqing-me/AttackVLM/assets/41732074/dc1abd4b-962b-4534-a17d-f0444f65347d">

These are the results for white-box attack using the above code. I am evaluating on 5000 images from ImageNet-1k with the frist 5000 captions from coco_captions.txt. I am able to get similar results to what are reported in the paper (Table 1).

Furthermore, regarding evaluating on BLIP model. Since the model takes input of size 384x384, while the adversarial examples are crafted on CLIP with a size of 224x224. Are the adversarial images resized to 384 for evaluation??

**Comment by HashmatShadab** [comment]
```
Batch 0
--------------------------------------------------
Average CLIP Score between adv_image and tgt_text: 0.653412401676178
Average CLIP Score between original image and tgt_text: 0.10163450241088867
Average CLIP Score between adv_text and tgt_text: 0.469323068857193
Average CLIP Score between clean_text and tgt_text: 0.5141135454177856
--------------------------------------------------
Batch 1
--------------------------------------------------
Average CLIP Score between adv_image and tgt_text: 0.5834584534168243
Average CLIP Score between original image and tgt_text: 0.09944814443588257
Average CLIP Score between adv_text and tgt_text: 0.5033599585294724
Average CLIP Score between clean_text and tgt_text: 0.588655561208725
--------------------------------------------------
Batch 2
--------------------------------------------------
Average CLIP Score between adv_image and tgt_text: 0.5613496899604797
Average CLIP Score between original image and tgt_text: 0.10016324122746785
Average CLIP Score between adv_text and tgt_text: 0.4908251365025838
Average CLIP Score between clean_text and tgt_text: 0.5873693029085795
--------------------------------------------------
Batch 3
--------------------------------------------------
Average CLIP Score between adv_image and tgt_text: 0.5778457820415497
Average CLIP Score between original image and tgt_text: 0.08570456225425005
Average CLIP Score between adv_text and tgt_text: 0.3918103128671646
Average CLIP Score between clean_text and tgt_text: 0.5153686329722404
--------------------------------------------------
```
During each batch, I am also evaluating the CLIP scores for Table 1 and Table 2. As you can see from the evaluation for the above batches, the attack is able to succesfully craft adversarial examples(white-box results), but the captions generated by BLIP-2 don't result in reproducing results for Table 2.

Kindly look into this and let me know if there is something missing in my code.

**Comment by yunqing-me** [comment]
Thank you for your interests. In general, you can use any open-sourced encoders like CLIP series (RN/ViT) for the transference. Therefore, when the model resolution is 384, there is no need to do any resize ops when you use clip model with 384 res (and similarly, for model input with 224 resolution, we use the same image encoder with 224 px). In this case, we used the same clip encoder as blip2 to transfer the white-box attack.

**Comment by klea-ziu** [comment]
I am facing similar problem when I try to reproduce the values of Table 2. I am using 10k data from MS-COCO and the results are lower. When I use 1000 data the results are similar for Unidiffuser MF-ii + MF-tt. Why is this happening?
